### ISSUE TITLE
Trade is sorely missing

### ISSUE THREAD

**Issue by Mutos** [naev/naev#849]
There is a sore miss in NAEV, player-wise : trading. The commodities exist and are displayed on planets, but they are at the same price everywhere.

I know the commerce and price computing model isn't officially fixed and there are a number of ways it could be designed, so for now, there is no way to make a "default" consistent model.

But the new player discovering NAEV doesn't know of that. They come to an EV-like game, and their basic expectation for such a game includes to be able to make money by trading stuff. So they land on a planet, buy some commodities and fly to another planet they feel fit to sell it. And bang, at that point, they discover that, wherever they land, it's always the same price.

That definitely can throw potentially interested players out of NAEV. It's about the very definition of the NAEV gameplay. We DO PROMISE trade, so delivering it is NOT AN OPTION, while honing it the way we think it should be under the hood DEFINITELY IS.  

So IMHO, it should be a highest-level priority to have at least a tentative, simple model : at least, something  that makes prices be different on different planets...

It could be done by making available a LUA price API on planets and systems, then implementing on LUA the simplest possible model. In the meantime, we could keep on arguing on what should be an ideal model, and having the LUA price API could even help us to prototype it. Multiple prototypes could be implemented and tested, helping us to make a choice. 

Or even further, if the LUA model is not too CPU-expensive, the system could stay like that in the long run, with a choice of multiple pricing system in the options and a default one corresponding to the one chosen by the devs... But that would require a little more effort...

**Comment by bobbens** [comment]
The main issue is it is not clear what model to use at all. Something simple would be do like what original EV did and basically hardcode the prices on all the planets, then use some events to basically move things up and down locally. However, this is not dynamic, and once the player figures out an "optimal trade route" it is fairly grind-enabling. Furthermore, my main worry is that pushing all the economy design to the developers is not something that scales well and also needs balancing. However, it would be fairly simple to implement this. Prices would be assigned to each commodity, and the planet editor would allow setting that. Event API could be added to read/write prices and some common event framework would be written to basically handle the entire price smudging. Should also be fairly straight-forward to integrate with the news so the player can get an idea of what to do.

Other options were a whitepaper I found like 10+ years ago on simulating the economy as an electrical grid. Intensity is amount of trade, voltage is prices, planets and stocks act as capacitors, and trade routes/piracy are resistances. While the idea was fairly interesting, in my intents on implementing it, in never worked out quite too well. I'm still not certain if my implementation sucked or if the idea just does not work.

**Comment by smeagolthellama** [comment]
Since there is already information (not sure how processable) on whether a planet is industrial, mining, or agricultural, I think that having a stable-ish supply and demand *should be* a part of the world. Oolite and other games in the elite family do this, so there isn't really any reason not to. Grinding might not be a good thing design-wise, but it is realistic. Some places produce food, others need it. Some places produce cheap goods, others ~~need~~ want them.
However, grinding is also boring. Why go for slow, boring money (repetitive trade), when you could go for quick exciting money (missions, etc.)?

Implementing a quick thing to change the price might be a bit tricky, if the needed data aren't there (how rich is a planet, what does it produce/need), so I think I have a workaround:
I noticed that on planets where they don't produce food, there isn't any for sale. This applies to other commodities as well. This is a good indicator of whether there is any demand: If they aren't selling, they are probably buying. So, one could say that they are going to be willing to pay you somewhere between about 1.2 and 2 times what you paid for, giving you a nice tidy profit (unless pirates, but that could be adjusted for by looking at the local piracy rate, maybe with the neighbouring piracy rate factored in a bit as well).

Event-wise, I'm not sure if there is anything that would impact the economy. I haven't gotten that far in the game yet.

**Comment by smeagolthellama** [comment]
Another thought I had for an economic model would be the "Urgent need for [insert commodity here]. Please bring as much as you can. We will pay x $/ton" type of missions. The maths  are already there, but I don't know if it's a good idea to generalise this.

**Comment by bobbens** [comment]
> Event-wise, I'm not sure if there is anything that would impact the economy. I haven't gotten that far in the game yet.

By events I meant the Lua events, which are basically like missions that the player has no control. This can be used to implement temporal price fluctuations to fake an economy like done in the original EV game, in which price drop/increases would randomly happen for various reasons. This framework already exists and new functions to modify prices would only have to be added.

**Comment by Mutos** [comment]
Hello bobbens, hello smeagolthellama,

Thanks for your answers. Of course, bobbens, I understand that getting an up-to-par economy is not an easy way, and that's not my purpose on this thread. The idea is just to make a basic framework that would :
1/ Enable different prices on each planet for a commodity, and thus, open the road towards trade gameplay and an economy system,
3/ Enable events to change prices from the original, uniform prices, allowing for first crude opportunities for the player to do something with the commodities,
2/ Allow to build prototypes of working economy systems, thus allowing ideas like your electrical diagram to be tested, so work towards a suitable system can be started.

@bobbens : when you say that the framework already exists, I'll dive into the code and see what is actually done. If you have already more information, could you share them on the thread, so as to orient my search ? Thanks in advance.

@smeagolthellama : I think the "urgent need" mission type is a good idea, as at first, it could be a motive for the player to seek assets selling the commodity in question, then when economy is introduced, to seek assets selling them cheap... Motive for exploring is always good...

**Comment by Mutos** [comment]
Another thought : there once was an economy branch by BariumBlue and PR #94 in 2014, on which onpon4 worked. Also PR #150, which is an updated one. I haven't read the whole discussion, where I sure could find why it hasn't been accepted. I'll also do this soon, so as to make a synthetic presentation here.

What I would want is to understand why the prices hooks are not in place and what least-effort action is actually needed to remove the roadblock.

**Comment by bobbens** [comment]
> @bobbens : when you say that the framework already exists, I'll dive into the code and see what is actually done. If you have already more information, could you share them on the thread, so as to orient my search ? Thanks in advance.

The event framework exists, the price framework does not. Looking at src/space.h:96, we can see that the commodities have a fixed price and not a per-planet price. The straight-forward way of implementing is to add a double field for each commodity, which would represent the local value price (1.0 would be original value, 1.2 would be 20% more expensive, etc...).  Next this should be edited to load the commodities prices in src/space.c:1997 planet_parse. They could just default to 1.0 if not set, so you would only have to define outliers. Of course, another way of handling this would be to give planet "templates" which would define prices which would be easier than doing per-planet. This would allow you to say define this planet as agricultural and have the prices automatically adjust on that. You would still have to be careful to not have two planets near each other have very different prices or players could exploit that. Afterwards this would have to be added to the system edit interface to be easier to handle.

With regarding events, they already exist but there is no framework to change planet prices. To do this, you would have to add some Lua functionality for this to for example src/nlua_planet.c and make sure the events can use this functionality (should be fine by default). A then single event should be created to trigger on when player enters, and it should be made to handle as much of the random events as possible. You just have to be careful to store some sort of table with all the information and reapply the price changes all the time or when the player loads the game they will be reset.

Regarding PR #94 and #150, I don't actually remember much what went on with them. Feel free to try it and if you believe it is stable enough, it might be worth merging. I don't think I ever got around to looking into them properly.

**Comment by Mutos** [comment]
Thanks @bobbens for the information,

Basically, I don't think we should merge #150, because it's so controversial. The overall plan I'm thinking of after taking your comments into account is the following :

1/ Just C-implement a per-planet price structure with appropriate Lua hooks to manipulate it,

2/ Merge this, then use it to create Lua-powered prototypes of what we could want for the economy,

3/ If in the course of these devs, we need some extra planet data, like templates, types, parameters or so, C-implement them and their Lua-hooks on their own branch,

4/ Test all that and discuss on the prototypes' actual results, merging the relevant template/type/parameter branches along the way,

5/ Then, and this could be months or years from now, decide and C-implement the chosen economic system.

what do you think of this ?

**Comment by bobbens** [comment]
> What do you think of this ?

I think it sounds good and is an easy way to get it running fast. As long as speed is not an issue, I have no problem leaving as much as possible in Lua. Actually, I think that's the better approach, as it is easier to edit and handle.

As I mentioned before, I think the main issues that will be introduced is
1) How to display price information to the player? Do we add an economy view to the starmap so the player can view, create, and set trade routes?
2) How to make it easier to edit and maintain? This is probably going to be the main issue. Using templates can alleviate this to a certain extent. Maybe instead of having them be planet-specific do system-specific to make editing easier. Or even just add some sort of modifier like faction presence, that bleeds to nearby systems, so you can create a price fluctuation map which will automatically handle nearby planets/systems in an easier way. That actually sounds like a fairly decent approach.
3) How to extend this? Going all Lua might be a bit tricky, so some functionality should be added into the engine in the long run. Taking inspiration from the pull requests might also be a good idea.

tl;dr go for it ;)

**Comment by ids1024** [comment]
> I think it sounds good and is an easy way to get it running fast. As long as speed is not an issue, I have no problem leaving as much as possible in Lua.

I don't think speed should be too much of an issue, since it's not something that runs every frame. Though I guess arbitrarily complicated economic models simulating the whole universe could be problematic.

LuaJIT is also known for its quite exceptional performance, though currently we support Lua 5.1 as well.

**Comment by Mutos** [comment]
Hello all,

@bobbens : as for your 3 points, that'll have to be addressed after the base layer is up and running, thus thinked over while we do the first layer :
1/ Adding an economy view to the map should be good,
2/ Editing and maintaining is important, I see once we have done the first layer, a second, more global layer that will be added. Your template approach seems fine...
3/ Extension and generalization is also an important issue. I think we should test in Lua, with a distinct tree for each prototype and branches for features, particularly C-implemented ones. Then, as the protos evolve and are tested, we can merge mature successful features from all protos into an intregration proto, to get a consistent global picture, and when this integration/staging proto is OK, merge it into mainstream NAEV.

**Comment by agb32** [comment]
Hi,
Is anyone actively working on this?  If not, I might take a look...

**Comment by Mutos** [comment]
Hi agb32,

Seems no-one, as far as I know. This blind spot is there since a long time and the need is blatant ^-^ so if you can take a look, I'll be happy to help !

I think the 3 main comments that could be taken as specs are bobbens', ids1024's and mine on 11-12/09/2019, and maybe ask bobbens to describe more precisely what he thinks to include into templates.

I'm no more fluent enough in C to tackle this one in /src, although I'm the one who whimpered on it :( But I can help for some testing and coding in /dat, as Lua and XML still fall into my skills ^-^

Lots of namasté, I was despairing this would interest no-one and stay in limbo ^-^

**Comment by agb32** [comment]
One thing that I don't understand the rationale for - if an asset does not have a certain commodity in the store (e.g. food), then this is presumably because the asset does not produce enough, and only imports it.  Therefore, if I have some in my hold, I should be able to sell it.  But this does not seem to be the case - even if I have some, I have no way of selling it.  Would it be okay for me to change this, so that everything can be sold, but only exported things can be bought?

**Comment by BTAxis** [comment]
I think that's a bad idea. That way literally every trade route is as short as the next port of call over. It also doesn't make sense for every planet to buy everything.

**Comment by agb32** [comment]
Good point - I'll leave it as it is in that case.

**Comment by Mutos** [comment]
Hi agb32, hi BTAxis,

I think limiting sales to a predefined list may be the fastest way to go, but basically a bad idea. For me, you have it, there's no logical in-universe pretext to prevent you to sell it, except if it's locally outlawed. But how could you explain that, for instance, food is locally outlawed ?

BUT... we must prevent what BTAxis told and which is true... So I suggest to have something like a floor price for a commodity, that guarantees you won't gain any money on such a sale, and you'll likely end up losing some.

I know a floor price is technically as artificial as a preset list of commodities, but in-universe, it makes more sense and doesn't make the player feel hampered.

Anyway, it may not be the higher priority ^-^ The priority imho would be to make prices asset-dependant and give Lua access to them. It may or not have been already done, I don't remember, but looking for it should also be a good way to dive into the trading screen code ^-^

**Comment by BTAxis** [comment]
It depends a bit on how you see commodity trade working. If you think of it as a sort of free-for-all bazaar where you trade with individuals purely on a supply and demand basis, then yes, it follows that you should be able to find a market for almost anything, however bad it may be.

But it's also plausible that trade is much more regulated, where you do business through channels set up by the local authorities. In that scenario, it follows that you would be able to trade only those commodities that are officially sanctioned at that particular port. Anything else is simply refused.

Naev's current system, as well as most trading games' systems, model the latter, which I think makes sense. On the other hand, I understand the argument that it's tedious to not be able to get rid of whatever's in your hold, especially if it's rare.

Perhaps a middle ground could be to have a "black market" of sorts, where you can trade goods not officially traded, at some kind of disadvantage. Some games do this, such as Starsector.

**Comment by Mutos** [comment]
Didn't we already implement a black market system ? I vaguely remember having seen that in some asset xmls.

**Comment by agb32** [comment]
Yes, there is some sort of black market - someone can ask you to get some goods, and they'll pay a certain price.  But not really trade.

I've now got a simple deterministic model working, which seems reasonable.  Basically, the price varies as a function of time as combination of 2 sine waves, frequency of one is system dependent, and the other is asset (planet) dependent.  The mean cost of the commodity depends on things like asset type, population size, etc, and the frequency and variation scale depend on some of these, the image used etc - basically everything within the xml files describing the system and assets, and then weighted by neighbouring systems.
e.g. price(t) = meanprice + sysScale * sin ( t/sysT) + assetScale * sin( t/assetT)
where meanprice > sysScale > assetScale.

This would only need to be calculated for a given asset upon landing there (and would then stay fixed for the visit - which I think is good).  In terms of memory, it is only 5 parameters per asset, so not a lot - 20-40 bytes (float or double).  So, only start up time would be affected, and this would be minimal - well less than a second.

This seems to give reasonable possibilities.  Trade within a system yields poor rewards, 1 jump away is far better, and 2,3,4 are better still.  

I'll begin to put this into the C code, as a first step, so that then we at least have something where we can usefully trade between systems.  Once its in, we can of course tweak the model simply enough.  But it should be fairly easy to do.

After that, I'll add bells and whistles as time allows - Lua interface, perhaps metrics to help the player, i.e. average price they've seen on places they've visit (which would require us saving the time at which they landed on each asset), the cost they bought what they currently have in their hold for, etc.

In terms of selling goods that you can't buy at a system, what I thought there is that you can sell them for a low price, which is guaranteed never to make you a profit (i.e. equal to the lowest possible purchase price).  So the only reason to do it would be if you want to free up cargo space, and don't care about making a loss (e.g. perhaps because you spotted a really good bargin on something else!).

If anyone is interested, I can share the python code I used to check this.  Its a bit messy, and basically parses the xml files, and then displays the data in various ways, computes some trade routes etc.

Let me know what you think!

**Comment by PhoenixRiver** [comment]
TLDR; Why not use safe lanes to modify commodity prices since they
asolutely would were we modelling non-player actual trade traffic?

A thought suddenly occurs... one of the problems I've seen in other
trade systems is that the player's trades make no difference and while
this works at the start once then player is trading in all of the
stock that a planet produces of a certain good it makes no sense that
the player doesn't change the balance of trade; the second thing that
seems to make little sense is that the further goods are taken, the
more the profit... why would food that is transported across half the
galaxy be any different from food grown next door? If you're a desert
world you want lotsof food whereever it comes from, if you're on a
luxuriant green agricultural world you don't want food at all, you
probably want to sell it since you have too much. It also means a
smaller very fast cargo ship might well make more commercial sense
than a slow huge cargo hauler (or group of haulers) since taking goods
as far as possible could well be more worth it than large amounts of
the same cargo a short distance and small cargo ships already have the
advantage of being more immune to pirates.

Given that the new "safe lanes" exist, could these be used as a basis
for trade routes that exist outside the player. Essentially the safe
lanes are used to calculate a shortest path to where a planet would
obtain a certain good; at that point you could decide how fine grained
you wanted to make the prices, anything from:
Have categories of commodity that each planet type produces (we
already have planet types after all and I did a pretty good job of
correcting and spreading them out a few years back), each category has
a default commodity price to sell what it produces with there being a
default price galaxy-wide for commodities that a planet does not
produce - each planet only sells those thing that it produces but buys
everything. The distance by number of hops in safe lanes is used to
modify the basic price for each category of commodity (since each
commodity in a category will have the sme number of hops since a
planet produces every commodity in a category). Lack of safe lane
would revert to whatever the maximum hop count on the table was (maybe
five or six) on the assumption it was not safe to transport cargo
there. The players hold would be inspected to see which categories
they are carrying and the price of those categories, for example, have
their hop count reduced by one.
to:
Calculate a price for each commodity for each planet based on the
exact distance (via space lane) with a value for how much of that
commodity the planet produces and consumes and readjusting the price
by a delta (small) random amount and recording the new price after the
player has bought and sold their cargo (this would be time degraded
when they visit again to work out how impactful their trade was).
Planets with no safe lanes could use the actual distance calculated
via the distance to and from portals insystem with some modifier added
for the pirate presense in each system and how far is travelled in
each system. In future it might be possible to allow a player to
entirely buy out some rarer commodities of a producing planet and
thereby inflat the prices at planets dependent on that supply...

I should note the above two calculation systems are supposed to be
opposite extremes that produce a system that is not so dissimilar to
real trade in a developed country.

This price calculation would reduce the amount of data being kept
about commodities since each planet would not keep it's own prices,
possibly it would keep player caused changes for a short time but the
majority of data would be by category or galaxy-wide (kinda like the
stock market) and the primary change would be the minimum travel cost
for a planet to obtain the commodity.

Apologies for the wall of text and complete lack of code, hopefully it
is helpful, if only in showing what not to do,

River

**Comment by agb32** [comment]
Okay - as a first step, I've created a first pull request.
This has a functioning pseudo economy.  Prices are computed during initialisation, along with variation scales and time periods.  As discussed above, these are dependent on planet and solar system characteristics, and vary sinusoidally with time
Upon landing at a planet and visiting the commodities tab, the current price of the commodity is then calculated (a simple arithmetic operation with a couple of sine functions).  

Next steps will be to add Lua hooks, and do some of the other things above.  But for now, at least we have something that will allow people to trade and make a profit, and also is not static, so gives a more exciting feeling.

I'm doing it in these small steps, because if life returns to normal (i.e. COVID-19), I'll have far less time free, so might never get the next steps done.

Please play around and have a go... (and hopefully accept the pull request before something else breaks it!)

Enjoy!

**Comment by agb32** [comment]
Re: Safe lanes - I'll try to factor these into my next update.
I know that the player's trades currently make no difference to the economy as discussed above - but one thing at a time I think, otherwise we'll never get trade in!  We must have been waiting for at least 10 years to be able to trade (I have, certainly!), so even something simple like this is good...
